sdks/go/pkg/beam/io/databaseio/database.go

```diff
@@ -0,0 +1,220 @@
+func Query(s beam.Scope, driver, dsn, q string, t reflect.Type) beam.PCollection {
+	s = s.Scope(driver + ".Query")
+	return query(s, driver, dsn, q, t)
+}
+
+func query(s beam.Scope, driver, dsn, query string, t reflect.Type) beam.PCollection {
+	imp := beam.Impulse(s)
+	return beam.ParDo(s, &queryFn{Driver: driver, Dsn: dsn, Query: query, Type: beam.EncodedType{T: t}}, imp, beam.TypeDefinition{Var: beam.XType, T: t})
+}
+
+type queryFn struct {
+	// Project is the project
+	Driver string `json:"driver"`
+	// Project is the project
+	Dsn string `json:"dsn"`
+	// Table is the table identifier.
+	Query string `json:"query"`
+	// Type is the encoded schema type.
+	Type beam.EncodedType `json:"type"`
+}
+
+func (f *queryFn) ProcessElement(ctx context.Context, _ []byte, emit func(beam.X)) error {
+	//TODO move DB Open and Close to Setup and Teardown methods or StartBundle and FinishBundle
+	db, err := sql.Open(f.Driver, f.Dsn)
+	if err != nil {
+		return fmt.Errorf("failed to open database: %v, %v", f.Driver, err)
+	}
+	defer db.Close()
+	statement, err := db.PrepareContext(ctx, f.Query)
+	if err != nil {
+		return fmt.Errorf("failed to prepare query: %v, %v", f.Query, err)
+	}
+	defer statement.Close()
+	rows, err := statement.QueryContext(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to run query: %v, %v", f.Query, err)
+	}
+	defer rows.Close()
+	var mapper rowMapper
+	var columns []string
+	for rows.Next() {
+		reflectRow := reflect.New(f.Type.T)
+		row := reflectRow.Interface() // row : *T
+		if mapper == nil {
+			columns, err = rows.Columns()
+			if err != nil {
+				return err
+			}
+			columnsTypes, _ := rows.ColumnTypes()
+			if mapper, err = newQueryMapper(columns, columnsTypes, f.Type.T); err != nil {
+				return fmt.Errorf("failed to create rowValues mapper: %v", err)
+			}
+		}
+		rowValues, err := mapper(reflectRow)
+		if err != nil {
+			return err
+		}
+		err = rows.Scan(rowValues...)
+		if err != nil {
+			return fmt.Errorf("failed to scan %v, %v", f.Query, err)
+		}
+		if loader, ok := row.(MapLoader); ok {
+			asDereferenceSlice(rowValues)
+			loader.LoadMap(asMap(columns, rowValues))
+		} else if loader, ok := row.(SliceLoader); ok {
+			asDereferenceSlice(rowValues)
+			loader.LoadSlice(rowValues)
+		}
+		emit(reflect.ValueOf(row).Elem().Interface()) // emit(*row)
+	}
+	return nil
+}
+
+// Write writes the elements of the given PCollection<T> to bigquery, if columns left empty all table columns are used to insert into, otherwise selected
+func Write(s beam.Scope, driver, dsn, table string, columns []string, col beam.PCollection) {
+	t := col.Type().Type()
+	s = s.Scope(driver + ".Write")
+	pre := beam.AddFixedKey(s, col)
+	post := beam.GroupByKey(s, pre)
+	beam.ParDo0(s, &writeFn{Driver: driver, Dsn: dsn, Table: table, Columns: columns, BatchSize: writeRowLimit, Type: beam.EncodedType{T: t}}, post)
+}
+
+// WriteWithBatchSize writes the elements of the given PCollection<T> to bigquery. T is required, with batchSize
```

Comment:
It's not clear what "T is required" means in this context. With sinks/writes, the type can be inferred from the PCollection.

Wouldn't batchSize be the maximum number of elements to include in a single write, rather than something to do with the schema type?

Reply:
Looks like doc was outdated, corrected it, batchSize controls elements used in batch INSERT SQL, i.e INSERT INTO t(a, b, c) VALUES(x1, y1, z1), ..., (xN, yN, yN); Batching can speed up loading data greatly, but not all SQL database may support it, so controlling batch size gives some lever for various SQL vendors.